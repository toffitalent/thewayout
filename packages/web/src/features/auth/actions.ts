import { CreateUserRequest } from '@two/shared';
import { API } from '@app/api';
import { createApiThunk } from '@app/utils';

export const fetchUser = createApiThunk('auth/fetchUser', API.user.retrieve);
export const forgotPassword = createApiThunk('auth/forgotPassword', API.user.forgotPassword);

export const login = createApiThunk(
  'auth/login',
  async (credentials: Parameters<typeof API.login>[0], { dispatch }) => {
    await API.login(credentials);
    await dispatch(fetchUser());
  },
);

export const logout = createApiThunk('auth/logout', API.logout);

export const resetPassword = createApiThunk(
  'auth/resetPassword',
  async (
    { username, password, token }: { username: string; password: string; token: string },
    { dispatch },
  ) => {
    await API.user.resetPassword(password, token);
    return dispatch(login({ username, password }));
  },
);

export const signUp = createApiThunk(
  'auth/register',
  async (payload: CreateUserRequest, { dispatch }) => {
    await API.user.create(payload);
    return dispatch(login({ username: payload.email, password: payload.password }));
  },
);

export const updateUser = createApiThunk('auth/updateUser', API.user.update);
