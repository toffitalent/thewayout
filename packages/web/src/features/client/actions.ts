import { ListRspClientRequest } from '@two/shared';
import { API } from '@app/api';
import type { State } from '@app/store';
import { createApiThunk } from '@app/utils';

export const createClient = createApiThunk('client/create', API.clientProfile.create);

export const updateClient = createApiThunk('client/update', API.clientProfile.update);

export const listAllJobs = createApiThunk(
  'client/jobs',
  ({ forceReload }: { forceReload?: boolean }, { getState }) => {
    const { ids } = (getState() as State).clientJobs.jobsList;
    const lastId = [...ids].pop() as string;
    return API.job.list({
      ...(!forceReload && { before: lastId }),
    });
  },
);

export const retrieveJob = createApiThunk(
  'client/job',
  ({ jobId, since }: { jobId: string; since?: Date }) =>
    API.job.retrieve(jobId, {
      ...(since ? { since } : {}),
    }),
);

export const applyJob = createApiThunk(
  'client/jobApply',
  ({ jobId, questions }: { jobId: string; questions?: { [key: string]: string }[] }) =>
    API.job.apply(jobId, {
      ...(questions ? { questions } : {}),
    }),
);

export const listAppliedJobs = createApiThunk(
  'client/applications',
  ({ forceReload, userId }: { forceReload?: boolean; userId: string }, { getState }) => {
    const { ids } = (getState() as State).clientJobs.appliedJobs;
    const lastId = [...ids].pop() as string;
    return API.clientProfile.listJobs(userId, {
      ...(!forceReload && { before: lastId }),
    });
  },
);

export const listSuggestedJobs = createApiThunk('client/suggestedJobs', (clientId: string) =>
  API.clientProfile.listSuggestedJobs(clientId),
);

export const listRsp = createApiThunk(
  'rsp/list',
  ({ userId, ...rest }: ListRspClientRequest & { userId: string }) =>
    API.clientProfile.listRsp({ userId, ...rest }),
);

export const retrieveCaseManager = createApiThunk(
  'client/caseManager',
  API.clientProfile.getCaseManager,
);
