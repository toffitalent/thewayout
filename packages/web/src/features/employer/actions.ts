import { Job, JobApplicationStatus } from '@two/shared';
import { API } from '@app/api';
import type { State } from '@app/store';
import { createApiThunk } from '@app/utils';

export const createEmployer = createApiThunk('employer/create', API.employer.create);

export const updateEmployer = createApiThunk('employer/update', API.employer.update);

export const createJob = createApiThunk('employer/job/create', API.job.create);

export const listJobs = createApiThunk(
  'employer/jobs',
  ({ forceReload, userId }: { forceReload?: boolean; userId: string }, { getState }) => {
    const { ids } = (getState() as State).employerJobs.jobs;
    const lastId = [...ids].pop() as string;
    return API.employer.listJobs(userId, {
      ...(!forceReload && { before: lastId }),
    });
  },
);

export const updateListJobs = createApiThunk(
  'employer/jobs/listUpdated',
  ({ userId }: { userId: string }, { getState }) => {
    const { listUpdatedAt, ids } = (getState() as State).employerJobs.jobs;
    const firstId = ids[0] as string;
    const lastId = [...ids].pop() as string;
    return API.employer.listJobs(userId, {
      since: listUpdatedAt,
      after: firstId,
      before: lastId,
    });
  },
);

export const retrieveJob = createApiThunk(
  'employer/job',
  ({ jobId, since }: { jobId: string; since?: Date }) =>
    API.job.retrieve(jobId, {
      ...(since ? { since } : {}),
    }),
);

export const listApplications = createApiThunk(
  'employer/listApplications',
  ({ employerId, jobId }: { employerId: string; jobId: string }, { getState }) => {
    const { entities } = (getState() as State).employerJobs.jobsApplications;
    const page = entities[jobId]?.page;
    const isListEmpty = !entities[jobId]?.applications.length;

    return API.employer.listApplications(employerId, jobId, {
      page: isListEmpty ? 0 : page! + 1,
    });
  },
);

export const getClient = createApiThunk(
  'employer/getClient',
  ({ employerId, jobId, clientId }: { employerId: string; jobId: string; clientId: string }) =>
    API.employer.getClient(employerId, jobId, clientId, {}),
);

export const updateJob = createApiThunk(
  'employer/updateJob',
  ({ jobId, patch }: { jobId: string; patch: Partial<Job> }) => API.job.update(jobId, patch),
);

export const interviewClient = createApiThunk(
  'employer/interviewClient',
  ({
    employerId,
    jobId,
    applicationId,
  }: {
    employerId: string;
    jobId: string;
    applicationId: string;
  }) =>
    API.employer.changeApplicationStatus(
      employerId,
      jobId,
      applicationId,
      JobApplicationStatus.interview,
    ),
);

export const hireClient = createApiThunk(
  'employer/hireClient',
  ({
    employerId,
    jobId,
    applicationId,
  }: {
    employerId: string;
    jobId: string;
    applicationId: string;
  }) =>
    API.employer.changeApplicationStatus(
      employerId,
      jobId,
      applicationId,
      JobApplicationStatus.hired,
    ),
);

export const rejectClient = createApiThunk(
  'employer/rejectClient',
  ({
    employerId,
    jobId,
    applicationId,
  }: {
    employerId: string;
    jobId: string;
    applicationId: string;
  }) =>
    API.employer.changeApplicationStatus(
      employerId,
      jobId,
      applicationId,
      JobApplicationStatus.rejected,
    ),
);

export const notAFitClient = createApiThunk(
  'employer/notAFitClient',
  ({
    employerId,
    jobId,
    applicationId,
  }: {
    employerId: string;
    jobId: string;
    applicationId: string;
  }) =>
    API.employer.changeApplicationStatus(
      employerId,
      jobId,
      applicationId,
      JobApplicationStatus.notAFit,
    ),
);
