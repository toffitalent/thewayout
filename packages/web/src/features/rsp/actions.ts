import { ClientType, RspClientStatus } from '@two/shared';
import { API } from '@app/api';
import type { State } from '@app/store';
import { createApiThunk } from '@app/utils';

export const createRsp = createApiThunk('rsp/create', API.rsp.create);

export const updateRsp = createApiThunk('rsp/update', API.rsp.update);

export const createMember = createApiThunk('rsp/createMember', API.rsp.createMember);

export const inviteCaseManager = createApiThunk('rsp/inviteCaseManager', API.rsp.inviteCaseManager);

export const removeInvitation = createApiThunk('rsp/removeInvitation', API.rsp.removeInvitation);

export const removeCaseManager = createApiThunk('rsp/removeCaseManager', API.rsp.removeCaseManager);

export const listCaseManagers = createApiThunk(
  'rsp/caseManagers',
  ({ rspId }: { rspId: string }, { getState }) => {
    const { ids } = (getState() as State).rsp.caseManagers;
    const lastId = [...ids].pop() as string;
    return API.rsp.listCaseManagers(rspId, {
      before: lastId,
    });
  },
);

export const listInvitations = createApiThunk(
  'rsp/invitations',
  ({ rspId }: { rspId: string }, { getState }) => {
    const { ids } = (getState() as State).rsp.invitations;
    const lastId = [...ids].pop() as string;
    return API.rsp.listInvitations(rspId, {
      before: lastId,
    });
  },
);

export const listCaseManagersNames = createApiThunk(
  'rsp/listCaseManagersNames',
  API.rsp.listCaseManagersNames,
);

export const retrieveRspAccount = createApiThunk('rsp/retrieveRspAccount', API.rsp.retrieveAccount);

export const updateCaseManager = createApiThunk('rsp/updateCaseManager', API.rsp.updateCaseManager);

export const listClients = createApiThunk(
  'rsp/clients',
  ({ rspId }: { rspId: string }, { getState }) => {
    const { ids } = (getState() as State).rsp.clients;
    const lastId = [...ids].pop() as string;
    return API.rsp.listClients(rspId, {
      before: lastId,
    });
  },
);

export const listStatisticsClients = createApiThunk(
  'rsp/statisticsClients',
  (
    {
      rspId,
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      forceReload,
      ...clientFilters
    }: Partial<ClientType> & { rspId: string; forceReload?: boolean },
    { getState },
  ) => {
    const { ids } = (getState() as State).rsp.statisticsClients;
    const lastId = [...ids].pop() as string;
    return API.rsp.listClients(rspId, {
      ...clientFilters,
      before: lastId,
    });
  },
);

export const retrieveClient = createApiThunk('rsp/retrieveClient', API.rsp.retrieveClient);

export const declineClient = createApiThunk(
  'rsp/declineClient',
  ({ rspId, clientId }: { rspId: string; clientId: string }) =>
    API.rsp.changeClientStatus(rspId, clientId, RspClientStatus.closed),
);

export const acceptClient = createApiThunk(
  'rsp/acceptClient',
  ({ rspId, clientId }: { rspId: string; clientId: string }) =>
    API.rsp.changeClientStatus(rspId, clientId, RspClientStatus.active),
);

export const addClientNotes = createApiThunk('rsp/addClientNotes', API.rsp.addClientNotes);

export const assignCaseManager = createApiThunk(
  'rsp/assignCaseManager',
  ({
    rspId,
    clientId,
    caseManagerId,
  }: {
    rspId: string;
    clientId: string;
    caseManagerId: string;
  }) => API.rsp.assignCaseManager(rspId, clientId, caseManagerId),
);

export const getStatistics = createApiThunk('rsp/statistics', API.rsp.getStatistics);
