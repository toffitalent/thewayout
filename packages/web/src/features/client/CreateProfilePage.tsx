import { Box, Toast, Wizard } from '@disruptive-labs/ui';
import { useRouter } from 'next/router';
import { SEO } from '@app/components/SEO';
import { WizardLayout } from '@app/components/WizardLayout';
import { useAppDispatch, useAppSelector, useErrorIndicator } from '@app/hooks';
import { getClientDataFromWizard } from '@app/utils';
import { selectAuthUser } from '../auth';
import { createClient } from './actions';
import styles from './CreateProfilePage.module.scss';
import { steps } from './profileData';
import { WizardData } from './profileDataTypes';

export const CreateProfilePage = () => {
  const user = useAppSelector(selectAuthUser);
  const dispatch = useAppDispatch();
  const showError = useErrorIndicator();
  const router = useRouter();

  const handleSave = (data: WizardData) => {
    dispatch(createClient(getClientDataFromWizard(data, true)))
      .unwrap()
      .catch((e) => {
        showError(e);
      })
      .then(() => {
        Toast.success('Profile added!');
      });
  };

  if (!user) {
    return null;
  }
  if (user.client) {
    router.push('/client/');
    return null;
  }

  const lastStep = steps.pop();
  if (lastStep) {
    lastStep.nextStep = (data) => handleSave(data as WizardData);
    steps.push(lastStep);
  }

  return (
    <>
      <SEO title="Create Profile" description="" />
      <Box className={styles.wizardWrapper}>
        <Wizard steps={steps} layout={WizardLayout} />
      </Box>
    </>
  );
};
