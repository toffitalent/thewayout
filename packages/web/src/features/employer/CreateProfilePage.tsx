import { Box, Wizard, WizardStep } from '@disruptive-labs/ui';
import { useRouter } from 'next/router';
import { useCallback } from 'react';
import { EmployerCreateRequest as WizardData } from '@two/shared';
import { SEO } from '@app/components/SEO';
import { WizardLayout } from '@app/components/WizardLayout';
import { useAppDispatch, useAppSelector, useErrorIndicator } from '@app/hooks';
import { selectAuthUser } from '../auth';
import { createEmployer } from './actions';
import { ConfirmationStep } from './ConfirmationStep';
import styles from './CreatorProfile.module.scss';
import { steps } from './profileData';

export const CreateProfilePage = () => {
  const user = useAppSelector(selectAuthUser);
  const router = useRouter();
  const dispatch = useAppDispatch();
  const showError = useErrorIndicator();

  const handleSubmit = useCallback(
    (data: WizardData) => {
      if (user) {
        dispatch(createEmployer(data))
          .unwrap()
          .then(() => {
            router.push('/employer/create-profile/confirm/');
          })
          .catch((e) => {
            showError(e);
          });
      }
    },
    [dispatch, router, showError, user],
  );

  if (!user) {
    return null;
  }

  const confirmationStep: WizardStep = {
    id: 'confirm',
    title: 'Confirm your data',
    component: <ConfirmationStep onSubmit={handleSubmit} />,
    showBackButton: true,
  };

  return (
    <>
      <SEO title="Create Profile" description="" />
      <Box className={styles.wizardWrapper}>
        <Wizard steps={[...steps, confirmationStep]} layout={WizardLayout} />
      </Box>
    </>
  );
};
