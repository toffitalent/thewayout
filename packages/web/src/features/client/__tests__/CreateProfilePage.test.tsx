import userEvent from '@testing-library/user-event';
import { useRouter } from 'next/router';
import {
  ExperienceSkills,
  JusticeStatus,
  OffenseCategory,
  PersonalStrengths,
  ReferredBy,
  ReleasedAt,
  StateFederal,
  Support,
  TimeServed,
  VeteranBranchOfService,
  VeteranCampaign,
  VeteranOrJustice,
  VeteranRank,
  VeteranTypeDischarge,
} from '@two/shared';
import { clientProfile as text } from '@app/data/clientProfileText';
import { act, fireEvent, fixtures, mockState, mockThunks, render, waitFor } from '@test';
import * as actions from '../actions';
import { CreateProfilePage } from '../CreateProfilePage';

jest.mock('@app/hooks', () => ({
  ...jest.requireActual('@app/hooks'),
  useAppAuth: jest.fn(),
}));

const partialRequestData = {
  address: '111 Street',
  city: 'Los Angeles',
  experience: [ExperienceSkills.carpentry],
  personalStrengths: [
    'communicatesWell',
    'creative',
    'dealsWellWithChange',
    'detailOriented',
    'goodWithTechnology',
  ],
  postalCode: '90111',
  referredBy: ReferredBy.familyFriends,
  state: 'CA',
  support: [Support.employment],
  disability: undefined,
  ethnicity: undefined,
  age: undefined,
  gender: undefined,
  maritalStatus: undefined,
  orientation: undefined,
  phone: null,
  religion: undefined,
  veteranStatus: undefined,
  relativeExperience: [],
  rspId: fixtures.rsp.id,
  isNewRspMember: false,
};
const partialJusticeImpactedData = {
  expectedReleasedAt: undefined,
  facility: undefined,
  releasedCounty: undefined,
};

const fillStepsForAllClients = async (findByRole: any) => {
  // Phone number
  await act(async () => fireEvent.click(await findByRole('button', { name: 'Save & Next' })));

  // Address number
  await userEvent.type(await findByRole('textbox', { name: 'Your Address' }), '111 Street', {
    delay: 1,
  });
  await userEvent.type(await findByRole('textbox', { name: 'Town / City' }), 'Los Angeles', {
    delay: 1,
  });
  await userEvent.selectOptions(await findByRole('combobox'), 'California');
  await userEvent.type(await findByRole('textbox', { name: 'Postcode / Zipcode' }), '90111', {
    delay: 1,
  });
  await act(async () => fireEvent.click(await findByRole('button', { name: 'Next' })));

  // Support step
  await act(async () =>
    fireEvent.click(await findByRole('checkbox', { name: text[Support.employment] })),
  );
  await act(async () => fireEvent.click(await findByRole('button', { name: 'Next' })));

  // Rsp step
  await act(async () => fireEvent.click(await findByRole('radio', { name: fixtures.rsp.name })));
  await act(async () => fireEvent.click(await findByRole('button', { name: 'Next' })));

  // Personal strengths step
  await act(async () =>
    fireEvent.click(
      await findByRole('checkbox', { name: text[PersonalStrengths.communicatesWell] }),
    ),
  );
  await act(async () =>
    fireEvent.click(await findByRole('checkbox', { name: text[PersonalStrengths.creative] })),
  );
  await act(async () =>
    fireEvent.click(
      await findByRole('checkbox', { name: text[PersonalStrengths.dealsWellWithChange] }),
    ),
  );
  await act(async () =>
    fireEvent.click(await findByRole('checkbox', { name: text[PersonalStrengths.detailOriented] })),
  );
  await act(async () =>
    fireEvent.click(
      await findByRole('checkbox', { name: text[PersonalStrengths.goodWithTechnology] }),
    ),
  );
  await act(async () => fireEvent.click(await findByRole('button', { name: 'Next' })));

  // Skills step
  await act(async () =>
    fireEvent.click(await findByRole('checkbox', { name: text[ExperienceSkills.carpentry] })),
  );
  await act(async () => fireEvent.click(await findByRole('button', { name: 'Next' })));

  // Language step
  await act(async () => fireEvent.click(await findByRole('button', { name: 'Skip' })));

  // Experience step
  await act(async () => fireEvent.click(await findByRole('button', { name: 'Skip' })));

  // Personal Summary step
  await act(async () => fireEvent.click(await findByRole('button', { name: 'Skip' })));

  // Education step
  await act(async () => fireEvent.click(await findByRole('button', { name: 'Skip' })));

  // License step
  await act(async () => fireEvent.click(await findByRole('button', { name: 'Skip' })));

  // Diversity step
  await act(async () => fireEvent.click(await findByRole('button', { name: 'Next' })));

  // Referred by step
  await act(async () =>
    fireEvent.click(await findByRole('radio', { name: text[ReferredBy.familyFriends] })),
  );
  await act(async () => fireEvent.click(await findByRole('button', { name: 'Submit' })));
};

describe('CreateProfilePage', () => {
  beforeEach(() => {
    (useRouter as jest.Mock).mockImplementation(() => ({
      beforePopState: jest.fn(),
    }));
  });

  test('renders correctly', async () => {
    const { container } = await waitFor(() =>
      render(<CreateProfilePage />, {
        initialState: { ...mockState, auth: { user: { ...fixtures.user, client: undefined } } },
      }),
    );
    expect(container).toMatchSnapshot();
  });

  test('renders no content if no user', async () => {
    const { container } = render(<CreateProfilePage />, {
      initialState: { ...mockState, auth: { user: null } },
    });

    expect(container).toMatchSnapshot();
  });

  test("show offense steps if else than 'No Offense' chosen", async () => {
    const [createClient] = mockThunks(actions, ['createClient']);
    const { findByRole, queryByText, dispatch } = render(<CreateProfilePage />, {
      initialState: { ...mockState, auth: { user: { ...fixtures.user, client: undefined } } },
    });

    // Veteran or Justice Impacted step
    await act(async () =>
      fireEvent.click(
        await findByRole('checkbox', { name: text[VeteranOrJustice.justiceImpacted] }),
      ),
    );
    await act(async () => fireEvent.click(await findByRole('button', { name: 'Save & Next' })));

    // Justice status step
    await act(async () =>
      fireEvent.click(await findByRole('radio', { name: text[JusticeStatus.halfwayHouse] })),
    );
    await act(async () => fireEvent.click(await findByRole('button', { name: 'Next' })));

    expect(queryByText("What's your phone number?")).not.toBeInTheDocument();
    expect(queryByText('Nature Of Offense(s)')).toBeInTheDocument();

    // Offense step
    await act(async () =>
      fireEvent.click(await findByRole('checkbox', { name: text[OffenseCategory.burglary] })),
    );
    await act(async () => fireEvent.click(await findByRole('button', { name: 'Save & Next' })));

    // Time served step
    await act(async () =>
      fireEvent.click(await findByRole('radio', { name: text[TimeServed.timeServed01] })),
    );
    await act(async () => fireEvent.click(await findByRole('button', { name: 'Next' })));

    // Completed sentence step
    await act(async () =>
      fireEvent.click(await findByRole('radio', { name: text[ReleasedAt.releasedAt01] })),
    );
    await act(async () => fireEvent.click(await findByRole('button', { name: 'Next' })));

    // State or federal
    await act(async () =>
      fireEvent.click(await findByRole('radio', { name: text[StateFederal.federal] })),
    );
    await act(async () => fireEvent.click(await findByRole('button', { name: 'Next' })));

    await fillStepsForAllClients(findByRole);

    expect(createClient).toBeCalledWith({
      veteranOrJustice: [VeteranOrJustice.justiceImpacted],
      ...partialRequestData,
      ...partialJusticeImpactedData,
      justiceStatus: JusticeStatus.halfwayHouse,
      offense: [OffenseCategory.burglary],
      releasedAt: ReleasedAt.releasedAt01,
      sbn: undefined,
      sexualOffenderRegistry: undefined,
      stateOrFederal: StateFederal.federal,
      timeServed: TimeServed.timeServed01,
      education: [],
      license: [],
    });
    expect(dispatch).toBeCalled();
  }, 20000);

  test("skip offense steps if 'No Offense' chosen", async () => {
    const [createClient] = mockThunks(actions, ['createClient']);
    const { findByRole, queryByText, dispatch } = render(<CreateProfilePage />, {
      initialState: { ...mockState, auth: { user: { ...fixtures.user, client: undefined } } },
    });

    // Veteran or Justice Impacted step
    await act(async () =>
      fireEvent.click(
        await findByRole('checkbox', { name: text[VeteranOrJustice.justiceImpacted] }),
      ),
    );
    await act(async () => fireEvent.click(await findByRole('button', { name: 'Save & Next' })));

    // Justice status step
    await act(async () =>
      fireEvent.click(await findByRole('radio', { name: text[JusticeStatus.noOffense] })),
    );
    await act(async () => fireEvent.click(await findByRole('button', { name: 'Next' })));

    expect(queryByText('Nature Of Offense(s)')).not.toBeInTheDocument();
    expect(queryByText("What's your phone number?")).toBeInTheDocument();

    await fillStepsForAllClients(findByRole);

    expect(createClient).toBeCalledWith({
      veteranOrJustice: [VeteranOrJustice.justiceImpacted],
      ...partialRequestData,
      ...partialJusticeImpactedData,
      justiceStatus: JusticeStatus.noOffense,
      offense: [],
      releasedAt: null,
      sbn: undefined,
      sexualOffenderRegistry: undefined,
      stateOrFederal: null,
      timeServed: null,
      education: [],
      license: [],
    });
    expect(dispatch).toBeCalled();
  }, 20000);

  test("Skip 'Completed sentence' step if 'Currently Incarcerated' justice option chosen and automatically set to 'Not Completed", async () => {
    const [createClient] = mockThunks(actions, ['createClient']);
    const { findByRole, queryByText, getByLabelText, dispatch } = render(<CreateProfilePage />, {
      initialState: { ...mockState, auth: { user: { ...fixtures.user, client: undefined } } },
    });

    // Veteran or Justice Impacted step
    await act(async () =>
      fireEvent.click(
        await findByRole('checkbox', { name: text[VeteranOrJustice.justiceImpacted] }),
      ),
    );
    await act(async () => fireEvent.click(await findByRole('button', { name: 'Save & Next' })));

    // Justice status step
    await act(async () =>
      fireEvent.click(
        await findByRole('radio', { name: text[JusticeStatus.currentlyIncarcerated] }),
      ),
    );
    await act(async () => fireEvent.click(await findByRole('button', { name: 'Next' })));

    // Correctional Facility step
    await userEvent.selectOptions(
      await findByRole('combobox'),
      'Chippewa Valley Correctional Treatment Facility',
    );
    await act(async () => fireEvent.click(await findByRole('button', { name: 'Save & Next' })));
    // Expected Release Date step
    await userEvent.type(
      getByLabelText('Expected Release Date'),
      `${new Date().getFullYear() + 1}-10-10}`,
      {
        delay: 1,
      },
    );
    await act(async () => fireEvent.click(await findByRole('button', { name: 'Save & Next' })));

    // County step
    await userEvent.selectOptions(await findByRole('combobox'), 'Adams');
    await act(async () => fireEvent.click(await findByRole('button', { name: 'Save & Next' })));

    expect(queryByText('Nature Of Offense(s)')).toBeInTheDocument();

    // Offense step
    await act(async () =>
      fireEvent.click(await findByRole('checkbox', { name: text[OffenseCategory.burglary] })),
    );
    await act(async () => fireEvent.click(await findByRole('button', { name: 'Save & Next' })));

    // Time served step
    await act(async () =>
      fireEvent.click(await findByRole('radio', { name: text[TimeServed.timeServed01] })),
    );
    await act(async () => fireEvent.click(await findByRole('button', { name: 'Next' })));

    // Here skip Completed sentence step

    // State or federal
    await act(async () =>
      fireEvent.click(await findByRole('radio', { name: text[StateFederal.federal] })),
    );
    await act(async () => fireEvent.click(await findByRole('button', { name: 'Next' })));

    await fillStepsForAllClients(findByRole);

    expect(createClient).toBeCalledWith(
      expect.objectContaining({
        veteranOrJustice: [VeteranOrJustice.justiceImpacted],
        releasedAt: ReleasedAt.notCompleted,
        expectedReleasedAt: `${new Date().getFullYear() + 1}-10-10`,
        facility: 'Chippewa Valley Correctional Treatment Facility',
        releasedCounty: 'Adams',
      }),
    );
    expect(dispatch).toBeCalled();
  }, 20000);

  test("show veteran steps if 'Veteran' chosen on Veteran or Justice Impacted step", async () => {
    const [createClient] = mockThunks(actions, ['createClient']);
    const { findByRole, findAllByRole, queryByText, dispatch } = render(<CreateProfilePage />, {
      initialState: { ...mockState, auth: { user: { ...fixtures.user, client: undefined } } },
    });

    // Veteran or Justice Impacted step
    await act(async () =>
      fireEvent.click(await findByRole('checkbox', { name: text[VeteranOrJustice.veteran] })),
    );
    await act(async () => fireEvent.click(await findByRole('button', { name: 'Save & Next' })));

    expect(queryByText('Branch of Service')).toBeInTheDocument();

    // Branch of Service step
    await act(async () =>
      fireEvent.click(await findByRole('checkbox', { name: text[VeteranBranchOfService.usArmy] })),
    );
    await act(async () => fireEvent.click(await findByRole('button', { name: 'Save & Next' })));

    // Member of the Reserves or National Guard step
    await act(async () => fireEvent.click(await findByRole('radio', { name: 'No' })));
    await act(async () => fireEvent.click(await findByRole('button', { name: 'Save & Next' })));

    // Campaigns of Service step
    await act(async () =>
      fireEvent.click(await findByRole('checkbox', { name: text[VeteranCampaign.bosnia] })),
    );
    await act(async () => fireEvent.click(await findByRole('button', { name: 'Save & Next' })));

    // Dates of Service step
    await userEvent.selectOptions((await findAllByRole('combobox'))[0], 'January');
    await userEvent.selectOptions((await findAllByRole('combobox'))[1], '2000');
    await userEvent.selectOptions((await findAllByRole('combobox'))[2], 'January');
    await userEvent.selectOptions((await findAllByRole('combobox'))[3], '2005');
    await act(async () => fireEvent.click(await findByRole('button', { name: 'Save & Next' })));

    // Type of Discharge step
    await userEvent.selectOptions(await findByRole('combobox'), text[VeteranTypeDischarge.bcd]);
    await act(async () => fireEvent.click(await findByRole('button', { name: 'Save & Next' })));

    // Rank at Discharge step
    await userEvent.selectOptions(await findByRole('combobox'), text[VeteranRank.AB]);
    await act(async () => fireEvent.click(await findByRole('button', { name: 'Save & Next' })));

    // DD-214 step
    await act(async () => fireEvent.click(await findByRole('radio', { name: 'No' })));
    await act(async () => fireEvent.click(await findByRole('button', { name: 'Save & Next' })));

    await fillStepsForAllClients(findByRole);

    expect(createClient).toBeCalledWith({
      veteranOrJustice: [VeteranOrJustice.veteran],
      ...partialRequestData,
      veteranService: [VeteranBranchOfService.usArmy],
      veteranRank: VeteranRank.AB,
      veteranStartAt: '2000-1-01',
      veteranEndAt: '2005-1-01',
      veteranReservist: false,
      veteranCampaigns: [VeteranCampaign.bosnia],
      veteranTypeDischarge: VeteranTypeDischarge.bcd,
      veteranDd214: false,
      education: [],
      license: [],
    });
    expect(dispatch).toBeCalled();
  }, 20000);
});
