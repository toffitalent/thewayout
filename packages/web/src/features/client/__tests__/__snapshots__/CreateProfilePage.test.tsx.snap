// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateProfilePage renders correctly 1`] = `
<div>
  <div
    class="wizardWrapper"
  >
    <section
      class="layout"
    >
      <div
        class="container"
      >
        <header
          class="header"
        >
          <button
            aria-label="Previous"
            class="button sm outline c-grey disabled iconButton sm round back backHidden"
            disabled=""
            type="button"
          >
            <svg
              aria-hidden="true"
              class="icon"
              focusable="false"
              stroke-width="2"
              viewBox="0 0 24 24"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
                stroke="none"
              />
              <polyline
                points="15 6 9 12 15 18"
              />
            </svg>
          </button>
          <div
            aria-valuemax="100"
            aria-valuemin="0"
            aria-valuenow="0"
            class="progress md rounded trackColor progress"
            role="progressbar"
          >
            <div
              class="track bgcolor-primary"
              style="transform: translateX(-100%);"
            />
          </div>
          <h2
            class="text title"
          >
            Are You a Veteran or Justice Impacted Individual?
          </h2>
          <p
            class="text description"
          >
            Select all that apply.
          </p>
        </header>
        <form
          class="form"
        >
          <div
            class="control control fluid"
            role="group"
          >
            <fieldset
              class="choices inline"
            >
              <div>
                <div
                  aria-checked="false"
                  aria-disabled="false"
                  class="choicesItem"
                  role="checkbox"
                  tabindex="0"
                >
                  <span
                    class="choicesItemText"
                  >
                    Veteran
                  </span>
                  <span
                    class="choicesItemIcon"
                  >
                    <svg
                      focusable="false"
                      role="presentation"
                      viewBox="0 0 24 24"
                    >
                      <polygon
                        fill="currentColor"
                        points="9.3,19 3,12.8 5.3,10.4 9.3,14.3 18.7,5 21,7.3 "
                      />
                    </svg>
                  </span>
                </div>
                <div
                  aria-checked="false"
                  aria-disabled="false"
                  class="choicesItem"
                  role="checkbox"
                  tabindex="0"
                >
                  <span
                    class="choicesItemText"
                  >
                    Justice Impacted
                  </span>
                  <span
                    class="choicesItemIcon"
                  >
                    <svg
                      focusable="false"
                      role="presentation"
                      viewBox="0 0 24 24"
                    >
                      <polygon
                        fill="currentColor"
                        points="9.3,19 3,12.8 5.3,10.4 9.3,14.3 18.7,5 21,7.3 "
                      />
                    </svg>
                  </span>
                </div>
              </div>
            </fieldset>
          </div>
          <button
            class="button md solid c-primary disabled fluid submit"
            disabled=""
            type="submit"
          >
            Save & Next
          </button>
        </form>
      </div>
    </section>
  </div>
</div>
`;

exports[`CreateProfilePage renders no content if no user 1`] = `<div />`;
